EXTENSIONS.md

```diff
@@ -0,0 +1,236 @@
+
+For these reasons, annotations are omitted from this proposal's "MVP" (minimum viable product) and considered as a possible follow-on proposal.
+
+## Function decorators and annotations
+
+The `@logged` decorator from earlier would Just Work(TM) on a function, with function decorators!
+
+```js
+@logged
+@{x: "y"}
+function f() { }
+
+f();                        // prints logging information
+f[Symbol.annotations][0].x  // "y"
+```
+
+Function declarations with decorators or annotations are not hoisted. This is because it would be unintuitive to reorder the evaluation of the decorator or annotation expressions.
+
+Instead, functions with decorators or annotations are defined only when their declaration is reached. If they are used before they are defined, a ReferenceError is thrown, like classes. This ReferenceError condition is sometimes referred to as a "temporal dead zone" (TDZ). The TDZ risks unfortunate situations when refactoring, but at least those situations lead to easy-to-debug errors rather than the wrong function being run.
+
+Function decorator details:
+- First parameter: the function being decorated (or, whatever the next inner decorator returned)
+- Second parameter: a context object which just has `{ kind: 'function' }`
+- Return value: a new function, or undefined to keep the same function
+
+The inner binding of a function expression inside itself is in TDZ until all the function decorators run.
+
+## Parameter decorators and annotations
+
+A parameter decorator wraps the value of a function/method argument. It returns a function which does the wrapping.
+
+```js
+function dec(_, context) {
+  assert(context.kind === "parameter");
+  return arg => arg - 1;
+}
+
+function f(@dec @{x: "y"} arg) { return arg * 2 ; }
+
+f(5)  // 8
+f[Symbol.annotations].parameters[0].x  // "y"
+```
+
+Functions with parameters that are decorated or annotated are treated similarly to decorated/annotated functions: they are not hoisted, and are in TDZ until their definition is executed.
+
+Parameter decorator details:
+- First parameter: undefined
+- Second parameter: a context object which just has `{ kind: 'parameter' }`
+- Return value: a function which takes a parameter value and returns a new parameter value. The function is called with the `this` value that the surrounding function is called with.
+
+This example can be desugared as
+
+```js
+let decInstance = dec(undefined, {kind: "parameter"});
+
+function f(arg_) {
+  const arg = decInstance(arg_);
+  return arg * 2 ;
+}
+
+f[Symbol.annotations] = []
+f[Symbol.annotations].parameters = []
+f[Symbol.annotations].parameters[0] = {x: "y"};
```

Comment:
this triggers a prototype setter, so it might be better to show it as:
```suggestion
f[Symbol.annotations].parameters = [
  {x: "y"},
];
```
?

Reply:
I don't think there would be any prototype setters involved, but yes, this change makes sense. Also, we should do s/annotations/metadata/ everywhere (annotations don't add a separate symbol).